paint/fill.go

```diff
@@ -0,0 +1,160 @@
+package paint
+
+import (
+	"errors"
+	"github.com/anthonynsimon/bild/clone"
+	"github.com/anthonynsimon/bild/util"
+	"image"
+	"image/color"
+	"math"
+)
+
+const (
+	MaxDistance = 510.0
+)
+
+type Point struct {
+	X, Y                  int
+	MarkedFromBelow       bool
+	MarkedFromAbove       bool
+	PreviousFillEdgeLeft  int
+	PreviousFillEdgeRight int
+}
+
+//Starts a flood fill of fillColor in the image with a start point and a defined fuzz (percentage of max color distance tolerated)
+//Returns filled image and error
+func FloodFill(img image.Image, startingPoint image.Point, fillColor color.Color, fuzz float64) (*image.RGBA, error) {
+
+	var st util.Stack
+	var point Point
+	visited := make(map[int]bool)
+	im := clone.AsRGBA(img)
+
+	maxX := im.Bounds().Dx() - 1
+	maxY := im.Bounds().Dy() - 1
+	if startingPoint.X > maxX || startingPoint.X < 0 || startingPoint.Y > maxY || startingPoint.Y < 0 {
+		return nil, errors.New("Starting point not in bounds!")
+	}
+
+	fuzzSquared := math.Pow(MaxDistance*fuzz/100, 2)
+	matchColor := im.At(startingPoint.X, startingPoint.Y)
+
+	st.Push(Point{startingPoint.X, startingPoint.Y, true, true, 0, 0})
+	pointsRemaining := st.Len() > 0
+
+	for pointsRemaining {
+		point = st.Pop().(Point)
+		pixOffset := im.PixOffset(point.X, point.Y)
+
+		if false == visited[pixOffset] {
+
+			im.Set(point.X, point.Y, fillColor)
+			visited[pixOffset] = true
+
+			//fill left side
+			xpos := point.X
+			for {
+				xpos--
+				if xpos < 0 {
+					xpos = 0
+					break
+				}
+				pixOffset = im.PixOffset(xpos, point.Y)
+				if isColorMatch(im, pixOffset, matchColor, fuzzSquared) {
+					im.Set(xpos, point.Y, fillColor)
+					visited[pixOffset] = true
+				} else {
+					break
+				}
+			}
+
+			leftFillEdge := xpos - 1
+			if leftFillEdge < 0 {
+				leftFillEdge = 0
+			}
+
+			//fill right side
+			xpos = point.X
+			for {
+				xpos++
+				if xpos > maxX {
+					break
+				}
+
+				pixOffset = im.PixOffset(xpos, point.Y)
+				if isColorMatch(im, pixOffset, matchColor, fuzzSquared) {
+					im.Set(xpos, point.Y, fillColor)
+					visited[pixOffset] = true
+				} else {
+					break
+				}
+			}
+			rightFillEdge := xpos + 1
+			if rightFillEdge > maxX {
+				rightFillEdge = maxX
+			}
+
+			//skip every second check for pixels above and below
```

Comment:
I've added the no skip version, and made two benchmarks. The skip version is ~30% faster.
The additional branching i think beats the pushing to the stack and checking has the point been already visited for almost 50% more points. It's your call which one we keep, i am more for the version with the column skip.